compute/backend_bucket: Fix default behavior with `USE_ORIGIN_HEADERS`

wyardley opened the pull request:
Since the send_empty_value field breaks the use of "USE_ORIGIN_HEADERS" cache_mode, adding a custom encoder to remove those fields when necessary.
    
Cherry-picked from @Alan-pad's rework of a commit by @d10i in #7062 
    
Fixes https://github.com/hashicorp/terraform-provider-google/issues/10622
Fixes https://github.com/hashicorp/terraform-provider-google/issues/13939
Closes #7062

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [ ] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:bug
compute: fixed an issue with TTLs being sent when `USE_ORIGIN_HEADERS` is set
```

ScottSuarez commented:
/gcbrun

wyardley commented:
@ScottSuarez @megan07 not sure if there are any issues I need to work on, but let me know. Seems like there are now more checks, but they still want a `/gcbrun` from a collaborator?

ScottSuarez commented:
hmm, could you try to rebase the PR? Recent ci changes might've caused issues here.

wyardley commented:
@ScottSuarez updated

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-7588-old..auto-pr-7588) ( 2 files changed, 80 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-7588-old..auto-pr-7588) ( 2 files changed, 80 insertions(+))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-7588-old..auto-pr-7588) ( 3 files changed, 41 insertions(+), 3 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-7588-old..auto-pr-7588) ( 3 files changed, 41 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `2559` 
Passed tests `2279` 
Skipped tests: `275` 
Affected tests: `5` 

#### Action taken 
<details> <summary>Found 5 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccDataSourceGoogleFirebaseAndroidAppConfig|TestAccComputeBackendBucket_withCdnPolicy|TestAccBigtableAppProfile_bigtableAppProfileSingleclusterExample|TestAccApigeeEnvKeystoreAliasPkcs12_apigeeEnvKeystoreAliasPkcs12Example|TestAccApigeeKeystoresAliasesKeyCertFile_apigeeKeystoresAliasesKeyCertFileTestExample </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccDataSourceGoogleFirebaseAndroidAppConfig`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/d31063e1-1cff-4cb0-a434-ef4a8fb900d1/recording/TestAccDataSourceGoogleFirebaseAndroidAppConfig.log)]
`TestAccComputeBackendBucket_withCdnPolicy`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/d31063e1-1cff-4cb0-a434-ef4a8fb900d1/recording/TestAccComputeBackendBucket_withCdnPolicy.log)]
`TestAccBigtableAppProfile_bigtableAppProfileSingleclusterExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/d31063e1-1cff-4cb0-a434-ef4a8fb900d1/recording/TestAccBigtableAppProfile_bigtableAppProfileSingleclusterExample.log)]
`TestAccApigeeEnvKeystoreAliasPkcs12_apigeeEnvKeystoreAliasPkcs12Example`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/d31063e1-1cff-4cb0-a434-ef4a8fb900d1/recording/TestAccApigeeEnvKeystoreAliasPkcs12_apigeeEnvKeystoreAliasPkcs12Example.log)]
`TestAccApigeeKeystoresAliasesKeyCertFile_apigeeKeystoresAliasesKeyCertFileTestExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/d31063e1-1cff-4cb0-a434-ef4a8fb900d1/recording/TestAccApigeeKeystoresAliasesKeyCertFile_apigeeKeystoresAliasesKeyCertFileTestExample.log)] 

All tests passed
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/d31063e1-1cff-4cb0-a434-ef4a8fb900d1/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/d31063e1-1cff-4cb0-a434-ef4a8fb900d1/recording) for each test

ScottSuarez commented on mmv1/templates/terraform/encoders/compute_backend_bucket.go.erb:
```diff
@@ -0,0 +1,35 @@
+c, cdnPolicyOk := d.GetOk("cdn_policy")
+
+if !cdnPolicyOk {
+  return obj, nil
+}
```
This is so it only applies durring update.. makes sense

ScottSuarez commented on mmv1/templates/terraform/encoders/compute_backend_bucket.go.erb:
```diff
@@ -0,0 +1,35 @@
+
+cdnPolicy := cdnPolicies[0].(map[string]interface{})
+
+cacheMode := cdnPolicy["cache_mode"].(string)
```
shouldn't cacheMode be coming from `obj` not from the existing state? Default from `obj` and fallback from state?

What if someone is changing the cacheMode to `USE_ORIGIN_HEADERS`

We only want to remove these

ScottSuarez commented on mmv1/templates/terraform/encoders/compute_backend_bucket.go.erb:
```diff
@@ -0,0 +1,35 @@
+cdnPolicy := cdnPolicies[0].(map[string]interface{})
+
+cacheMode := cdnPolicy["cache_mode"].(string)
+_, defaultTTLOk := cdnPolicy["default_ttl"]
+_, clientTTLOk := cdnPolicy["client_ttl"]
```
```suggestion
_, defaultTTLOk := cdnPolicy["default_ttl"]
_, clientTTLOk := cdnPolicy["client_ttl"]
```
These seem irrelevant to me. It doesn't matter what is in state here. It only matters what is in the `obj` or what we are sending to the api.

ScottSuarez commented on mmv1/templates/terraform/encoders/compute_backend_bucket.go.erb:
```diff
@@ -0,0 +1,35 @@
+_, defaultTTLOk := cdnPolicy["default_ttl"]
+_, clientTTLOk := cdnPolicy["client_ttl"]
+
+if obj["cdnPolicy"] == nil {
+  obj["cdnPolicy"] = make(map[string]interface{})
+}
```
if obj is not sending cdnpolicy we can return here.  We only want to remove the values when they are present. If obj is not PATCHing them then we shouldn't do anything.

wyardley replied:
Will the resources be created with the correct (null) values initially if we only do something on `PATCH`?

ScottSuarez commented on mmv1/templates/terraform/encoders/compute_backend_bucket.go.erb:
```diff
@@ -0,0 +1,35 @@
+
+switch cacheMode {
+case "USE_ORIGIN_HEADERS":
+  if _, ok := encCDNPolicy["defaultTtl"]; ok && defaultTTLOk {
```
```suggestion
  if v, ok := encCDNPolicy["defaultTtl"]; ok && v == 0 {
```

ScottSuarez replied:
note should we throw an error if v is not 0? That would infer that the user is trying to set the value even though it's unsettable with `USE_ORIGIN_HEADERS`

```golang
v, ok := encCDNPolicy["defaultTtl"]
if ok && v!= 0 {
// error?
}
```

ScottSuarez reviewed (changes requested):
requesting some changes.. Also could you add comments inline to explain why the encoder is needed. This is a pretty esoteric issue so it would help for future readers.

wyardley commented:
@ScottSuarez I can take a look, but keep in mind, I’m not the original author, nor am I the person who last adapted this. I was taking a stab at addressing the feedback from #7062

wyardley commented:
@ScottSuarez going to build this locally, but is this closer to what you were looking for?

wyardley commented on mmv1/templates/terraform/encoders/compute_backend_bucket.go.erb:
```diff
@@ -0,0 +1,35 @@
+	return obj, nil
+}
+
+cdnPolicies, err := c.([]interface{})
```
@ScottSuarez This was my attempt at resolving this error with applying using the built provider:

```
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-google plugin:

panic: interface conversion: interface {} is nil, not string
```

so I'm thinking if this interface is nil, should be fine to return `obj, nil` there?

wyardley commented on mmv1/templates/terraform/encoders/compute_backend_bucket.go.erb:
```diff
@@ -0,0 +1,35 @@
+}
+
+// Don't do anything if we're not PATCHing
+if obj["cdnPolicy"] == nil {
```
is this how you were expecting to reference the object?

ScottSuarez replied:
This makes sense to me

wyardley commented:
@ScottSuarez so now I have something that applies, but still exhibits the problem in question (i.e., API is responding with this after create
```
2023-04-10T10:06:38.485-0700 [DEBUG] provider.terraform-provider-google:  "cdnPolicy": {
2023-04-10T10:06:38.485-0700 [DEBUG] provider.terraform-provider-google:   "cacheMode": "USE_ORIGIN_HEADERS",
2023-04-10T10:06:38.485-0700 [DEBUG] provider.terraform-provider-google:   "clientTtl": 0,
2023-04-10T10:06:38.485-0700 [DEBUG] provider.terraform-provider-google:   "defaultTtl": 0,
```

and throwing this error 
```
│ Error: Error updating BackendBucket "projects/xxxx/global/backendBuckets/xxxx": googleapi: Error 400: Invalid value for field 'resource.cdnPolicy.defaultTtl': '0'. default_ttl cannot be specified with USE_ORIGIN_HEADERS cache_mode., invalid
```

So I'm thinking `c.([]interface{})` is coming back `nil` even when it shouldn't?

ScottSuarez commented:
> @ScottSuarez I can take a look, but keep in mind, I’m not the original author, nor am I the person who last adapted this. I was taking a stab at addressing the feedback from #7062

Hey @wyardley, it's difficult for my to also gain context and test this thoroughly. Albeit looking into this deeply will be equivalent in term of simply implementing this fix myself. Our team is pretty stretched as is.

wyardley commented:
So, the good news is that I’m setup / able to test more easily now. The bad news is that with all these changes, it doesn’t seem to actually fix the problem.

I am going to see if the original code (before requested changes) works / fixes the issue, and if so, I’ll start over with that as a starting point.

wyardley commented:
@ScottSuarez we can see how tests look in CI, but the current version I think addresses most / all of your feedback, and, more importantly, seems to work in my local testing.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-7588-old..auto-pr-7588) ( 2 files changed, 91 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-7588-old..auto-pr-7588) ( 2 files changed, 91 insertions(+))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-7588-old..auto-pr-7588) ( 3 files changed, 52 insertions(+), 3 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-7588-old..auto-pr-7588) ( 3 files changed, 41 insertions(+), 3 deletions(-))

ScottSuarez commented on mmv1/templates/terraform/encoders/compute_backend_bucket.go.erb:
```diff
@@ -0,0 +1,46 @@
+// level error)
+c, cdnPolicyOk := d.GetOk("cdn_policy")
+
+// Only apply during update
+if !cdnPolicyOk {
+	return obj, nil
+}
+
+// Don't do anything if we're not PATCHing
+if obj["cdnPolicy"] == nil {
+	return obj, nil
+}
```
nit: condense if statements

ScottSuarez commented on mmv1/templates/terraform/encoders/compute_backend_bucket.go.erb:
```diff
@@ -0,0 +1,46 @@
+	return obj, nil
+}
+
+// object is not sending cdnPolicy, so we can return early here as well
+if len(cdnPolicies) == 0 {
+	return obj, nil
+}
```
You're checking cndPolicy from state here. Not the cdnPolicy from obj.

ScottSuarez commented on mmv1/templates/terraform/encoders/compute_backend_bucket.go.erb:
```diff
@@ -0,0 +1,46 @@
+	return obj, nil
+}
+
+cdnPolicies := c.([]interface{})
```
Nit: it might make sense to rephrase these as currentCDNPolicies and futureCDNPolicies. It's difficult to understand the logic implicitly otherwise.

wyardley replied:
done

ScottSuarez reviewed (changes requested):
This makes sense to me. Asking for some code cleanup

modular-magician commented:
#### Tests analytics 
Total tests: `2564` 
Passed tests `2286` 
Skipped tests: `275` 
Affected tests: `3` 

#### Action taken 
<details> <summary>Found 3 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccDataSourceGoogleFirebaseAndroidAppConfig|TestAccApigeeKeystoresAliasesKeyCertFile_apigeeKeystoresAliasesKeyCertFileTestExample|TestAccApigeeKeystoresAliasesPkcs12_ApigeeKeystoresAliasesPkcs12Example </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccDataSourceGoogleFirebaseAndroidAppConfig`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/920478c8-e839-4318-9851-d55c44cb45ac/recording/TestAccDataSourceGoogleFirebaseAndroidAppConfig.log)]
`TestAccApigeeKeystoresAliasesKeyCertFile_apigeeKeystoresAliasesKeyCertFileTestExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/920478c8-e839-4318-9851-d55c44cb45ac/recording/TestAccApigeeKeystoresAliasesKeyCertFile_apigeeKeystoresAliasesKeyCertFileTestExample.log)]
`TestAccApigeeKeystoresAliasesPkcs12_ApigeeKeystoresAliasesPkcs12Example`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/920478c8-e839-4318-9851-d55c44cb45ac/recording/TestAccApigeeKeystoresAliasesPkcs12_ApigeeKeystoresAliasesPkcs12Example.log)] 

All tests passed
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/920478c8-e839-4318-9851-d55c44cb45ac/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/920478c8-e839-4318-9851-d55c44cb45ac/recording) for each test

wyardley commented:
Thanks, will push some updates based on the feedback in a few, after testing again. Let me know if I got it right in terms of current vs. future - I agree that it will make this much more readable. Appreciate the quick review, there's a lot involved in getting environment setup to test this well locally.

q: does magic modules have a position on whether the indenting in the erb templates should follow ruby or golang style? It seems somewhat inconsistent in the codebase. I've been using hard-tabs, which is what I've done on previous PRs, though I know it will get processed by `go fmt` either way.

Updated version seems to work for the happy path, but I am able to get it into a wacky state by first setting `cache_mode` to `"CACHE_ALL_STATIC"` with `max_ttl` set to a value, and then back to `"USE_ORIGIN_HEADERS"` with `max_ttl` undefined. We may need to not let the perfect be the enemy of the good in terms of fixing this bug (and have people delete / reimport the resource if they need to switch modes, though if you have any thoughts about potential fix, happy to try and test (setting `max_ttl` to `null` explicitly also doesn't work here).

```
│ Error: Error updating BackendBucket "projects/xxx/global/backendBuckets/yyy": googleapi: Error 400: Invalid value for field 'resource.cdnPolicy.maxTtl': '600'. max_ttl must be specified with CACHE_ALL_STATIC cache_mode only., invalid
│ 
│   with module.deleteme.google_compute_backend_bucket.cdn_backend_bucket,
│   on ../git/terraform/modules/gcp/lb_cdn_bucket/loadbalancer.tf line 16, in resource "google_compute_backend_bucket" "cdn_backend_bucket":
│   16: resource "google_compute_backend_bucket" "cdn_backend_bucket" {
```

I notice that in this case, it seems to be a `PUT`, so maybe we need to remove the bit about only doing it when we're doing a `PATCH` operation? Going to try commenting out that early return to see if that helps.

```
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google: PUT /compute/v1/projects/xxx/global/backendBuckets/yyyy?alt=json HTTP/1.1
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google: Host: compute.googleapis.com
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google: User-Agent: Terraform/1.4.4 (+https://www.terraform.io) Terraform-Plugin-SDK/2.10.1 terraform-provider-google/dev
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google: Content-Length: 513
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google: Content-Type: application/json
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google: X-Goog-User-Project: xxx
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google: Accept-Encoding: gzip
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google: 
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google: {
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google:  "bucketName": "deleteme-xyz",
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google:  "cdnPolicy": {
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google:   "cacheMode": "USE_ORIGIN_HEADERS",
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google:   "maxTtl": 300,
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google:   "negativeCaching": false,
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google:   "requestCoalescing": false,
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google:   "serveWhileStale": 0,
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google:   "signedUrlCacheMaxAgeSec": 0
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google:  },
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google:  "customResponseHeaders": [
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google:   "Permissions-Policy: camera=(), geolocation=*, microphone=()",
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google:   "Referrer-Policy: no-referrer-when-downgrade",
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google:   "Strict-Transport-Security: max-age=2592000; includeSubDomains",
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google:   "X-Content-Type-Options: nosniff",
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google:   "X-Frame-Options: SAMEORIGIN"
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google:  ],
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google:  "enableCdn": true,
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google:  "name": "deleteme-xxx-backend"
2023-04-11T14:58:53.202-0700 [DEBUG] provider.terraform-provider-google: }
```

wyardley commented:
@ScottSuarez Ok -- I updated it to include max TTL as well, and works for me locally as expected with no further changes. I think I've incorporated your requested feedback as well. Thanks for all the help / suggestions.

Q: will the existing test cases already be testing switching between modes, or do we need to explicitly test that case as well?

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-7588-old..auto-pr-7588) ( 2 files changed, 85 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-7588-old..auto-pr-7588) ( 2 files changed, 85 insertions(+))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-7588-old..auto-pr-7588) ( 3 files changed, 46 insertions(+), 3 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-7588-old..auto-pr-7588) ( 3 files changed, 41 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `2609` 
Passed tests `2328` 
Skipped tests: `277` 
Affected tests: `4` 

#### Action taken 
<details> <summary>Found 4 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccFirebaserulesRelease_BasicRelease|TestAccDataSourceGoogleFirebaseAndroidAppConfig|TestAccApigeeKeystoresAliasesPkcs12_ApigeeKeystoresAliasesPkcs12Example|TestAccApigeeKeystoresAliasesKeyCertFile_apigeeKeystoresAliasesKeyCertFileTestExample </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccFirebaserulesRelease_BasicRelease`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/5a1421af-34f6-4365-b57e-0dd550f74252/recording/TestAccFirebaserulesRelease_BasicRelease.log)]
`TestAccApigeeKeystoresAliasesPkcs12_ApigeeKeystoresAliasesPkcs12Example`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/5a1421af-34f6-4365-b57e-0dd550f74252/recording/TestAccApigeeKeystoresAliasesPkcs12_ApigeeKeystoresAliasesPkcs12Example.log)]
`TestAccApigeeKeystoresAliasesKeyCertFile_apigeeKeystoresAliasesKeyCertFileTestExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/5a1421af-34f6-4365-b57e-0dd550f74252/recording/TestAccApigeeKeystoresAliasesKeyCertFile_apigeeKeystoresAliasesKeyCertFileTestExample.log)] 

Tests failed during RECORDING mode:
 `TestAccDataSourceGoogleFirebaseAndroidAppConfig`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/5a1421af-34f6-4365-b57e-0dd550f74252/build-log/recording_build/TestAccDataSourceGoogleFirebaseAndroidAppConfig_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/5a1421af-34f6-4365-b57e-0dd550f74252/recording/TestAccDataSourceGoogleFirebaseAndroidAppConfig.log)] 

Please fix these to complete your PR
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/5a1421af-34f6-4365-b57e-0dd550f74252/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/5a1421af-34f6-4365-b57e-0dd550f74252/recording) for each test

wyardley commented:
Looks like the failing tests are unrelated?

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-7588-old..auto-pr-7588) ( 2 files changed, 85 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-7588-old..auto-pr-7588) ( 2 files changed, 85 insertions(+))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-7588-old..auto-pr-7588) ( 3 files changed, 46 insertions(+), 3 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-7588-old..auto-pr-7588) ( 3 files changed, 41 insertions(+), 3 deletions(-))

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccApigeeKeystoresAliasesKeyCertFile_apigeeKeystoresAliasesKeyCertFileTestExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/96af6ada-c047-4ae2-9525-d63becaa1e31/recording/TestAccApigeeKeystoresAliasesKeyCertFile_apigeeKeystoresAliasesKeyCertFileTestExample.log)]
`TestAccApigeeKeystoresAliasesPkcs12_ApigeeKeystoresAliasesPkcs12Example`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/96af6ada-c047-4ae2-9525-d63becaa1e31/recording/TestAccApigeeKeystoresAliasesPkcs12_ApigeeKeystoresAliasesPkcs12Example.log)]
`TestAccDataSourceGoogleFirebaseAndroidAppConfig`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/96af6ada-c047-4ae2-9525-d63becaa1e31/recording/TestAccDataSourceGoogleFirebaseAndroidAppConfig.log)] 

All tests passed
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/96af6ada-c047-4ae2-9525-d63becaa1e31/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-7588/artifacts/96af6ada-c047-4ae2-9525-d63becaa1e31/recording) for each test
